File: 2014 Southeast USA Regionals Division 1/goldleaf.pas
program SymmetryAxes;

type
  Tuple = record
    r1, c1, r2, c2: Integer;
  end;

  TupleArray = array of Tuple;

function CompareTuples(const a, b: Tuple): Integer;
begin
  if a.r1 <> b.r1 then
    CompareTuples := a.r1 - b.r1
  else if a.c1 <> b.c1 then
    CompareTuples := a.c1 - b.c1
  else if a.r2 <> b.r2 then
    CompareTuples := a.r2 - b.r2
  else
    CompareTuples := a.c2 - b.c2;
end;

procedure AddTuple(var arr: TupleArray; r1, c1, r2, c2: Integer);
var
  len: Integer;
begin
  len := Length(arr);
  SetLength(arr, len + 1);
  arr[len].r1 := r1;
  arr[len].c1 := c1;
  arr[len].r2 := r2;
  arr[len].c2 := c2;
end;

function ValidVert(r1, c1, r2, c2, n, m: Integer; grid: array of string): Boolean;
var
  i, j, i1, j1: Integer;
begin
  for i := 0 to n - 1 do
  begin
    for j := 0 to m - 1 do
    begin
      i1 := i;
      j1 := 2 * c1 - j + 1;
      if (i1 < 0) or (i1 >= n) or (j1 < 0) or (j1 >= m) then
      begin
        if grid[i][j+1] = '.' then
        begin
          ValidVert := False;
          Exit;
        end
        else
          Continue;
      end;
      if grid[i][j+1] = grid[i1][j1+1] then
      begin
        ValidVert := False;
        Exit;
      end;
    end;
  end;
  ValidVert := True;
end;

procedure GetVert(n, m: Integer; grid: array of string; var rv: TupleArray);
var
  i: Integer;
begin
  for i := 0 to m - 1 do
  begin
    if ValidVert(0, i, n-1, i, n, m, grid) then
    begin
      AddTuple(rv, 0, i, n-1, i);
      Exit;
    end;
  end;
end;

function ValidHor(r1, c1, r2, c2, n, m: Integer; grid: array of string): Boolean;
var
  i, j, i1, j1: Integer;
begin
  for i := 0 to n - 1 do
  begin
    for j := 0 to m - 1 do
    begin
      i1 := 2 * r1 - i + 1;
      j1 := j;
      if (i1 < 0) or (i1 >= n) or (j1 < 0) or (j1 >= m) then
      begin
        if grid[i][j+1] = '.' then
        begin
          ValidHor := False;
          Exit;
        end
        else
          Continue;
      end;
      if grid[i][j+1] = grid[i1][j1+1] then
      begin
        ValidHor := False;
        Exit;
      end;
    end;
  end;
  ValidHor := True;
end;

procedure GetHor(n, m: Integer; grid: array of string; var rv: TupleArray);
var
  i: Integer;
begin
  for i := 0 to n - 1 do
  begin
    if ValidHor(i, 0, i, m-1, n, m, grid) then
    begin
      AddTuple(rv, i, 0, i, m-1);
      Exit;
    end;
  end;
end;

function ValidDiag1(r1, c1, r2, c2, n, m: Integer; grid: array of string): Boolean;
var
  i, j, i1, j1, s: Integer;
begin
  s := r1 + c1;
  for i := 0 to n - 1 do
  begin
    for j := 0 to m - 1 do
    begin
      i1 := s - j;
      j1 := s - i;
      if (i1 < 0) or (i1 >= n) or (j1 < 0) or (j1 >= m) then
      begin
        if grid[i][j+1] = '.' then
        begin
          ValidDiag1 := False;
          Exit;
        end
        else
          Continue;
      end;
      if (i = i1) and (j = j1) then
      begin
        if grid[i][j+1] = '.' then
        begin
          ValidDiag1 := False;
          Exit;
        end
        else
          Continue;
      end;
      if grid[i][j+1] = grid[i1][j1+1] then
      begin
        ValidDiag1 := False;
        Exit;
      end;
    end;
  end;
  ValidDiag1 := True;
end;

procedure GetDiag1(n, m: Integer; grid: array of string; var rv: TupleArray);
var
  i, j, r1, c1, r2, c2: Integer;
begin
  for i := 0 to n - 1 do
  begin
    for j := 0 to m - 1 do
    begin
      if (j <> 0) and (i <> n - 1) then
        Continue;
      r1 := i; c1 := j;
      r2 := i; c2 := j;
      while (r2 >= 0) and (c2 < m) do
      begin
        Dec(r2); Inc(c2);
      end;
      Inc(r2); Dec(c2);
      if ValidDiag1(r1, c1, r2, c2, n, m, grid) then
      begin
        AddTuple(rv, r1, c1, r2, c2);
        Exit;
      end;
    end;
  end;
end;

function ValidDiag2(r1, c1, r2, c2, n, m: Integer; grid: array of string): Boolean;
var
  i, j, i1, j1, s: Integer;
begin
  s := r1 - c1;
  for i := 0 to n - 1 do
  begin
    for j := 0 to m - 1 do
    begin
      i1 := s + j;
      j1 := i - s;
      if (i1 < 0) or (i1 >= n) or (j1 < 0) or (j1 >= m) then
      begin
        if grid[i][j+1] = '.' then
        begin
          ValidDiag2 := False;
          Exit;
        end
        else
          Continue;
      end;
      if (i = i1) and (j = j1) then
      begin
        if grid[i][j+1] = '.' then
        begin
          ValidDiag2 := False;
          Exit;
        end
        else
          Continue;
      end;
      if grid[i][j+1] = grid[i1][j1+1] then
      begin
        ValidDiag2 := False;
        Exit;
      end;
    end;
  end;
  ValidDiag2 := True;
end;

procedure GetDiag2(n, m: Integer; grid: array of string; var rv: TupleArray);
var
  i, j, r1, c1, r2, c2: Integer;
begin
  for i := 0 to n - 1 do
  begin
    for j := 0 to m - 1 do
    begin
      if (j <> 0) and (i <> 0) then
        Continue;
      r1 := i; c1 := j;
      r2 := i; c2 := j;
      while (r2 < n) and (c2 < m) do
      begin
        Inc(r2); Inc(c2);
      end;
      Dec(r2); Dec(c2);
      if ValidDiag2(r1, c1, r2, c2, n, m, grid) then
      begin
        AddTuple(rv, r1, c1, r2, c2);
        Exit;
      end;
    end;
  end;
end;

procedure SortTuples(var arr: TupleArray);
var
  i, j: Integer;
  temp: Tuple;
begin
  for i := 0 to High(arr) - 1 do
  begin
    for j := i + 1 to High(arr) do
    begin
      if CompareTuples(arr[i], arr[j]) > 0 then
      begin
        temp := arr[i];
        arr[i] := arr[j];
        arr[j] := temp;
      end;
    end;
  end;
end;

var
  n, m, i: Integer;
  grid: array of string;
  rv: TupleArray;
begin
  ReadLn(n, m);
  SetLength(grid, n);
  for i := 0 to n - 1 do
    ReadLn(grid[i]);
  
  SetLength(rv, 0);
  GetVert(n, m, grid, rv);
  GetHor(n, m, grid, rv);
  GetDiag1(n, m, grid, rv);
  GetDiag2(n, m, grid, rv);
  
  if Length(rv) > 0 then
  begin
    SortTuples(rv);
    with rv[0] do
      WriteLn(r1 + 1, ' ', c1 + 1, ' ', r2 + 1, ' ', c2 + 1);
  end;
end.

File: 2014 Southeast USA Regionals Division 1/goldleaf.java
import java.util.*;

public class goldleaf {
    static class Tuple implements Comparable<Tuple> {
        int r1, c1, r2, c2;
        
        Tuple(int r1, int c1, int r2, int c2) {
            this.r1 = r1;
            this.c1 = c1;
            this.r2 = r2;
            this.c2 = c2;
        }
        
        @Override
        public int compareTo(Tuple other) {
            if (r1 != other.r1) return Integer.compare(r1, other.r1);
            if (c1 != other.c1) return Integer.compare(c1, other.c1);
            if (r2 != other.r2) return Integer.compare(r2, other.r2);
            return Integer.compare(c2, other.c2);
        }
    }
    
    static boolean validVert(int r1, int c1, int r2, int c2, int n, int m, char[][] grid) {
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                int i1 = i;
                int j1 = 2 * c1 - j + 1;
                if (i1 < 0 || i1 >= n || j1 < 0 || j1 >= m) {
                    if (grid[i][j] == '.') {
                        return false;
                    } else {
                        continue;
                    }
                }
                if (grid[i][j] == grid[i1][j1]) {
                    return false;
                }
            }
        }
        return true;
    }
    
    static void getVert(int n, int m, char[][] grid, TreeSet<Tuple> rv) {
        for (int i = 0; i < m; i++) {
            int r1 = 0, c1 = i, r2 = n - 1, c2 = i;
            if (validVert(r1, c1, r2, c2, n, m, grid)) {
                rv.add(new Tuple(r1, c1, r2, c2));
                return;
            }
        }
    }
    
    static boolean validHor(int r1, int c1, int r2, int c2, int n, int m, char[][] grid) {
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                int i1 = 2 * r1 - i + 1, j1 = j;
                if (i1 < 0 || i1 >= n || j1 < 0 || j1 >= m) {
                    if (grid[i][j] == '.') {
                        return false;
                    } else {
                        continue;
                    }
                }
                if (grid[i][j] == grid[i1][j1]) {
                    return false;
                }
            }
        }
        return true;
    }
    
    static void getHor(int n, int m, char[][] grid, TreeSet<Tuple> rv) {
        for (int i = 0; i < n; i++) {
            int r1 = i, c1 = 0, r2 = i, c2 = m - 1;
            if (validHor(r1, c1, r2, c2, n, m, grid)) {
                rv.add(new Tuple(r1, c1, r2, c2));
                return;
            }
        }
    }
    
    static boolean validDiag1(int r1, int c1, int r2, int c2, int n, int m, char[][] grid) {
        int s = r1 + c1;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                int i1 = s - j, j1 = s - i;
                if (i1 < 0 || i1 >= n || j1 < 0 || j1 >= m) {
                    if (grid[i][j] == '.') {
                        return false;
                    } else {
                        continue;
                    }
                }
                if (i == i1 && j == j1) {
                    if (grid[i][j] == '.') {
                        return false;
                    } else {
                        continue;
                    }
                }
                if (grid[i][j] == grid[i1][j1]) {
                    return false;
                }
            }
        }
        return true;
    }
    
    static void getDiag1(int n, int m, char[][] grid, TreeSet<Tuple> rv) {
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (j != 0 && i != n - 1) {
                    continue;
                }
                int r1 = i, c1 = j, r2 = i, c2 = j;
                while (r2 >= 0 && c2 < m) {
                    r2--; c2++;
                }
                r2++; c2--;
                if (validDiag1(r1, c1, r2, c2, n, m, grid)) {
                    rv.add(new Tuple(r1, c1, r2, c2));
                    return;
                }
            }
        }
    }
    
    static boolean validDiag2(int r1, int c1, int r2, int c2, int n, int m, char[][] grid) {
        int s = r1 - c1;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                int i1 = s + j, j1 = i - s;
                if (i1 < 0 || i1 >= n || j1 < 0 || j1 >= m) {
                    if (grid[i][j] == '.') {
                        return false;
                    } else {
                        continue;
                    }
                }
                if (i == i1 && j == j1) {
                    if (grid[i][j] == '.') {
                        return false;
                    } else {
                        continue;
                    }
                }
                if (grid[i][j] == grid[i1][j1]) {
                    return false;
                }
            }
        }
        return true;
    }
    
    static void getDiag2(int n, int m, char[][] grid, TreeSet<Tuple> rv) {
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (j != 0 && i != 0) {
                    continue;
                }
                int r1 = i, c1 = j, r2 = i, c2 = j;
                while (r2 < n && c2 < m) {
                    r2++; c2++;
                }
                r2--; c2--;
                if (validDiag2(r1, c1, r2, c2, n, m, grid)) {
                    rv.add(new Tuple(r1, c1, r2, c2));
                    return;
                }
            }
        }
    }
    
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int m = scanner.nextInt();
        scanner.nextLine(); // consume newline
        
        char[][] grid = new char[n][m];
        for (int i = 0; i < n; i++) {
            String line = scanner.nextLine();
            grid[i] = line.toCharArray();
        }
        
        TreeSet<Tuple> rv = new TreeSet<>();
        getVert(n, m, grid, rv);
        getHor(n, m, grid, rv);
        getDiag1(n, m, grid, rv);
        getDiag2(n, m, grid, rv);
        
        if (!rv.isEmpty()) {
            Tuple first = rv.first();
            System.out.println((first.r1 + 1) + " " + (first.c1 + 1) + " " + 
                             (first.r2 + 1) + " " + (first.c2 + 1));
        }
    }
}